core/src/main/java/com/yahoo/oak/OakHashMap.java

```diff
@@ -0,0 +1,580 @@
+/*
+ * Copyright 2020, Verizon Media.
+ * Licensed under the terms of the Apache 2.0 license.
+ * Please see LICENSE file in the project root for terms.
+ */
+
+package com.yahoo.oak;
+
+import java.util.AbstractCollection;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * A concurrent map implementation which supports off-heap memory.
+ */
+public class OakHashMap<K, V> extends AbstractMap<K, V> implements AutoCloseable, ConcurrentMap<K , V> {
+
+//    private final InternalOakHashMap<K , V> internalOakHashMap;
+    /*
+     * Memory manager cares for allocation, de-allocation and reuse of the internally pre-allocated
+     * memory. There can be separate memory managing algorithms for keys and values.
+     * */
+    private final MemoryManager valuesMemoryManager;
+    private final MemoryManager keysMemoryManager;
+    private final OakTransformer<K> keyDeserializeTransformer;
+    private final OakTransformer<V> valueDeserializeTransformer;
+    private final Function<Map.Entry<OakScopedReadBuffer, OakScopedReadBuffer>,
+            Map.Entry<K, V>> entryDeserializeTransformer;
+    private final OakComparator<K> comparator;
+    private final InternalOakMap<K , V> internalOakHashMap;
+
+
+
+    private final K fromKey;
+    private final boolean fromInclusive;
+    private final K toKey;
+    private boolean toInclusive;
+    private final boolean isDescending;
+    // internal constructor, to create OakHashMap use OakMapBuilder
+    OakHashMap(K minKey,
+               OakSerializer<K> keySerializer,
+               OakSerializer<V> valueSerializer,
+               OakComparator<K> oakComparator,
+               int chunkMaxItems,
+               MemoryManager vMM,
+               MemoryManager kMM) {
+        this.valuesMemoryManager = vMM;
+        this.keysMemoryManager = kMM;
+        this.comparator = oakComparator;
+        this.keyDeserializeTransformer = keySerializer::deserialize;
+        this.valueDeserializeTransformer = valueSerializer::deserialize;
+        this.entryDeserializeTransformer = entry -> new AbstractMap.SimpleEntry<>(
+                keySerializer.deserialize(entry.getKey()),
+                valueSerializer.deserialize(entry.getValue()));
+        this.internalOakHashMap = new InternalOakMap<>(minKey, keySerializer, valueSerializer, oakComparator,
+                this.valuesMemoryManager, kMM, chunkMaxItems, new ValueUtils());
+
+
+        this.fromKey = null;
```

Comment:
Thinking deeper about it, assigning all those parameters to undefined value, making it impossible to do parametrized testing for iterators... @liran-funaro what do you think to rewrite iterator's invocation code for OakHash inside ConcurrentZCMap interface?